plugins/okta/main.go

```diff
@@ -0,0 +1,434 @@
+	case "okta.security.asnumber":
+		req.SetValue(data.SecurityContext.AsNumber)
+	case "okta.security.asorg":
+		req.SetValue(data.SecurityContext.AsOrg)
+	case "okta.security.isp":
+		req.SetValue(data.SecurityContext.ISP)
+	case "okta.security.domain":
+		req.SetValue(data.SecurityContext.Domain)
+	case "okta.target.user.id":
+		for _, i := range data.Target {
+			if i.Type == "User" {
+				req.SetValue(i.ID)
+			}
+		}
+	case "okta.target.user.alternateid":
+		for _, i := range data.Target {
+			if i.Type == "User" {
+				req.SetValue(i.AlternateID)
+			}
+		}
+	case "okta.target.user.name":
+		for _, i := range data.Target {
+			if i.Type == "User" {
+				req.SetValue(i.DisplayName)
+			}
+		}
+	case "okta.target.group.id":
+		for _, i := range data.Target {
+			if i.Type == "UserGroup" {
+				req.SetValue(i.ID)
+			}
+		}
+	case "okta.target.group.alternateid":
+		for _, i := range data.Target {
+			if i.Type == "UserGroup" {
+				req.SetValue(i.AlternateID)
+			}
+		}
+	case "okta.target.group.name":
+		for _, i := range data.Target {
+			if i.Type == "UserGroup" {
+				req.SetValue(i.DisplayName)
+			}
+		}
+	// case "okta.security.isproxy":
+	// 	req.SetValue(data.SecurityContext.IsProxy)
+	default:
+		return fmt.Errorf("no known field: %s", req.Field())
+	}
+
+	return nil
+}
+
+// Open is called by Falco plugin framework for opening a stream of events, we call that an instance
+func (oktaPlugin *OktaPlugin) Open(params string) (source.Instance, error) {
+	req, err := http.NewRequest("GET", fmt.Sprintf("https://%v.%v", oktaPlugin.Organization, oktaBaseURL), nil)
+	if err != nil {
+		return &OktaInstance{}, err
+	}
+
+	since := time.Now().UTC().Add(-60 * time.Second)
+	values := req.URL.Query()
+	values.Add("since", since.Format(time.RFC3339))
+	values.Add("limit", "1000")
+	req.URL.RawQuery = values.Encode()
+
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Authorization", "SSWS "+oktaPlugin.APIToken)
+
+	return &OktaInstance{
+		client:  &http.Client{},
+		request: req,
+		ctx:     context.Background(),
+	}, nil
+}
+
+// String represents the raw value of on event
+// (not currently used by Falco plugin framework, only there for future usage)
+func (oktaPlugin *OktaPlugin) String(in io.ReadSeeker) (string, error) {
+	evtBytes, err := ioutil.ReadAll(in)
+	if err != nil {
+		return "", err
+	}
+	evtStr := string(evtBytes)
+
+	return fmt.Sprintf("%v", evtStr), nil
+}
+
+// NextBatch is called by Falco plugin framework to get a batch of events from the instance
+func (oktaInstance *OktaInstance) NextBatch(pState sdk.PluginState, evts sdk.EventWriters) (int, error) {
+	var logEvents []LogEvent
+
+	resp, err := oktaInstance.client.Do(oktaInstance.request)
+	if err != nil {
+		return 0, err
+	}
+	defer resp.Body.Close()
+
+	err = json.NewDecoder(resp.Body).Decode(&logEvents)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(logEvents) == 0 {
+		time.Sleep(5 * time.Second)
```

Comment:
For avoiding to reach rate limit. The last timestamp is already stored by updating the query parameter in the request
`values.Set("since", t.Add(1*time.Second).Format(time.RFC3339))`